File: DataStructure/Search&Sort/quick_sort.cpp
//
// Created by 薛智钧 on 2020/4/5.
//
#include <cstdio>
#include <cstdlib>
#include <algorithm>

int partition(int arr[], int lo, int hi){
    std::swap(arr[lo], arr[lo + rand() % (hi - lo)]); // 随机交换
    hi--;
    int pivot = arr[lo];
    while (lo < hi){
        while (lo < hi && pivot <= arr[hi]) --hi; // 向左拓展G
        if (lo < hi) arr[lo++] = arr[hi]; // 小于pivot者，归入左侧
        while (lo < hi && arr[lo] <= pivot) ++lo; // 向右拓展L
        if (lo < hi) arr[hi--] = arr[lo]; // 大于pivot者，归入右侧
    }
    arr[lo] = pivot;
    return lo;
}

void quickSort(int arr[], int lo, int hi){
    if (hi - lo < 2) return;

    int mi = partition(arr, lo, hi); // 获取轴点位置
    quickSort(arr, lo, mi); // 对轴点左侧进行排序
    quickSort(arr, mi + 1, hi);  // 对轴点右侧进行排序
}

void printArray(int arr[], int len){
    int i = 0;
    for(; i < len - 1; ++i)
        printf("%d, ", arr[i]);
    printf("%d", arr[i]);
    printf("\n");
}

int main(){
    int arr[] = {3,9,7,5,6,8,8};
    int len = sizeof(arr) / sizeof(arr[0]);

    printf("The given array is: \n");
    printArray(arr, len);
    quickSort(arr, 0, len);
    printf("The sorted result is: \n");
    printArray(arr, len);
    return 0;
}

File: DataStructure/Search&Sort/quick_sort.py
import random 

def qsort(arr):
    n = len(arr)
    _qsort(arr, 0, n-1)

def _qsort(arr, lo, hi):
    if lo < hi:
        p = partition(arr, lo, hi)
        _qsort(arr, lo, p-1)
        _qsort(arr, p+1, hi)

def partition(arr, lo, hi):
    random.shuffle(arr[lo : hi+1])
    pivot = arr[lo]
    while lo < hi:
        while lo < hi and pivot <= arr[hi]: hi -= 1
        if lo < hi: 
            arr[lo] = arr[hi]
            lo += 1
        while lo < hi and arr[lo] <= pivot: lo += 1
        if lo < hi:
            arr[hi] = arr[lo]
            hi -= 1
    arr[lo] = pivot # important
    return lo

if __name__ == "__main__":
    a = [2,34,5,-1,9,10]
    qsort(a)
    print(a)